### DISCUSSION TITLE
【0x02 Project】使用 eth.build 调用智能合约

### DISCUSSION THREAD

**Discussion by leeduckgo** [NonceGeek/Move-Camp#6] [Projects]
> Project 0x02 建议完成时长：2h

在上一个例子里，我们使用 `eth.build` 学习了如何发送 ETH，同时借助辅助资料了解到 ETH 中的基础 —— Transaction。

在今天的 Project 0x02 中，我们来实践如何通过`eth.build`来进行智能合约的调取！

## 0x01 智能合约的部署

Remix 是和 Eth.build 并行的代码沙盒工具，Eth.build 主要职能是帮助我们通过无代码的形式理解以太坊的原理，Remix 则可以让我们编写与调用智能合约！

![image-20211109185214101](https://tva1.sinaimg.cn/large/008i3skNgy1gw93gb0j5wj311o0u0dic.jpg)

> https://remix.ethereum.org/

### 1.1 新建智能合约

打开`Remix`，新建智能合约：

```solidity
contract HelloWorld {
    string public test = "hello world";
}
```

![image-20211109190435583](https://tva1.sinaimg.cn/large/008i3skNgy1gw93sjmummj31e90u00wp.jpg)

### 1.2 部署智能合约

切换到 `Deploy` 标签，选择`Injected Web3` 环境，连接 Metamask（注意 Metamask选择 Ropsten 测试网络）。

![image-20211109190533363](https://tva1.sinaimg.cn/large/008i3skNgy1gw93tjn68hj31e60u0n0p.jpg)

如果没有 `Ropsten` 的测试币，通过下述水龙头地址拿币：

> https://faucet.ropsten.be/

![image-20211109190821858](https://tva1.sinaimg.cn/large/008i3skNgy1gw93wh0uijj327w0t8tc7.jpg)

部署合约后拿到合约地址：

![image-20211109190846058](https://tva1.sinaimg.cn/large/008i3skNgy1gw93wvllwlj30ia06674d.jpg)

这是一个已经在`Ropsten`上部署好的合约地址：

> 0x7CF156A3eed0fae30385AB5dB4bCbf356FaF0e98

## 0x02 通过 eth.build 调用智能合约

这是写好的一个例子：

> https://eth.build/build#64e5e4416cdfbccddafdc69c0c92c581f86bcfcdc8cbeefb494f0a3dd4dc0690

这里我们可以填写我们在0x01中部署好的智能合约。

![image-20211109191122905](https://tva1.sinaimg.cn/large/008i3skNgy1gw93zlmgtuj30xk0ao3zf.jpg)

要注意 Chain 中的网络要和我们所部署的智能合约网络一致，此处都是 Ropsten 网络即可。

![image-20211109191240405](https://tva1.sinaimg.cn/large/008i3skNgy1gw940xxycrj30ly09cmxk.jpg)

Abi通过Compile组件（`web3` -> `compile` ）生成。

![image (3)](https://tva1.sinaimg.cn/large/008i3skNgy1gw94xl29iaj306e030a9y.jpg)

compile使用 0x01 中部署的智能合约及合约名字作为输入。

![image (4)](https://tva1.sinaimg.cn/large/008i3skNgy1gw94ylmezjj30pm095q3o.jpg)

使用compile输出的abi作为contract的abi输入，此时我们可以通过test()方法获取智能合约中变量的值。

![image-20211109194540408](https://tva1.sinaimg.cn/large/008i3skNgy1gw94zayjbwj315m0okjtn.jpg)

## 0x03 补充资料

本 Projects 的资料主要是智能合约相关：

- [微芒专题 | 14天区块链极速入门](https://ke.qq.com/course/3614841)

![image-20211109191734139](https://tva1.sinaimg.cn/large/008i3skNgy1gw9461gxrnj31bo0but9e.jpg)

- [CryptoZombies](https://cryptozombies.io/)
- [如何理解以太坊ABI - 应用程序二进制接口](https://learnblockchain.cn/2018/08/09/understand-abi)
- [跟我学 Solidity ：开发环境](https://learnblockchain.cn/article/1755)

## 0x04 挑战

试试其他智能合约吧？可以从网络上找找有什么可以玩的简单合约！

**Comment by xiaoyue2019** [comment]
https://eth.build/build#a4e9f4195c3cecdcf2e35d87ff10ad75a883c6e94d1da657f0f88e9285ef5092

**Comment by lysion** [comment]
https://sandbox.eth.build/build#9c3c19c01a30e0360023ddb02d860571cdce450671b5ed0171e8ad1a9afe3ced

**Comment by lysion** [comment]
1.1 新建智能合约
打开Remix，新建智能合约：

contract HelloWorld {
    string public **_tet_** = "hello world";
}

这里应该是少了个s

**Reply by leeduckgo** [reply]
感谢指出~

**Comment by sunliying2020** [comment]
https://eth.build/build#af21b53b906d5919695bd28264ccf2d88133c4b4fb0af4f74dfb839375e82c5c

**Reply by sunliying2020** [reply]
https://shimo.im/docs/TCcwcygwKhWTtkdJ/ 《eth.build学习笔记》

**Comment by hechty** [comment]
完成

https://eth.build/build#4fdc377ef370249148f573d0f7d10b316252acd8b9b639f4c37c8804a4228ae2

**Reply by leeduckgo** [reply]
不错👍

**Comment by gofenix** [comment]
https://eth.build/build#0c5bd5b982b27257e606d45700a52673228f8face9888c8353a1de94c307a6f4

**Comment by xFANx** [comment]
https://eth.build/build#2ebf77df7b27bcddeadfdb6a7c36d140c5594e69da0a2ef9a9365a6a2a81a90c
